server/engine/src/main/java/org/eclipse/lsp/cobol/dialects/ibm/TransformTreeStage.java

```diff
@@ -134,20 +133,101 @@ protected List<Node> transformAST(AnalysisContext ctx,
         new CobolVisitor(copybooksRepository, tokens, ctx.getExtendedDocument(),
             messageService, subroutineService, cachingConfigurationService, cobolProgramLayout);
     List<Node> syntaxTree = visitor.visit(tree);
+    shapeSectionsAndParagraphs(syntaxTree.get(0));
     ctx.getAccumulatedErrors().addAll(visitor.getErrors());
     return syntaxTree;
   }
 
-  private Node processSyntaxTree(AnalysisConfig analysisConfig, SymbolAccumulatorService symbolAccumulatorService, AnalysisContext ctx, List<Node> syntaxTree) {
-    Node rootNode = syntaxTree.get(0);
+  private void shapeSectionsAndParagraphs(Node parent) {
+    LinkedList<Node> stack = new LinkedList<>();
+    for (Node node : parent.getChildren()) {
+      parent.removeChild(node);
+      if (!node.getChildren().isEmpty()) {
+          shapeSectionsAndParagraphs(node);
+      }
+      if (!stack.isEmpty() && !RangeUtils.isInside(node.getLocality().getRange(), stack.peek().getLocality().getRange())) {
```

Comment:
What does `node.getLocality().getRange()` return if a node originates from a copybook?

Reply:
Good point. I've found a better way to resolve it :)